src/crawlee/storages/_storage_instance_manager.py

```diff
@@ -1,21 +1,49 @@
 from __future__ import annotations
 
-from collections.abc import Awaitable, Callable
+from collections import defaultdict
+from collections.abc import Coroutine, Hashable
+from dataclasses import dataclass, field
 from typing import TYPE_CHECKING, TypeVar
 
 from crawlee.storage_clients._base import DatasetClient, KeyValueStoreClient, RequestQueueClient
 
-from ._base import Storage
-
 if TYPE_CHECKING:
-    from crawlee.configuration import Configuration
+    from crawlee.storage_clients import StorageClient
+
+    from ._base import Storage
 
 T = TypeVar('T', bound='Storage')
 
+
+@dataclass
+class _StorageClientCache:
+    """Cache for specific storage client.
+
+    Example:
+        Storage=Dataset, id='123', additional_cache_key="some_path" will be located in
+        storage =  by_id[Dataset]['123'][some_path]
+    """
+
+    by_id: defaultdict[type[Storage], defaultdict[str, defaultdict[Hashable, Storage]]] = field(
+        default_factory=lambda: defaultdict(lambda: defaultdict(lambda: defaultdict()))
+    )
+    """Cache for storage instances by ID, separated by storage type and additional hash key."""
+
+    by_name: defaultdict[type[Storage], defaultdict[str, defaultdict[Hashable, Storage]]] = field(
+        default_factory=lambda: defaultdict(lambda: defaultdict(lambda: defaultdict()))
+    )
+    """Cache for storage instances by name, separated by storage type and additional hash key."""
+
+    by_alias: defaultdict[type[Storage], defaultdict[str, defaultdict[Hashable, Storage]]] = field(
+        default_factory=lambda: defaultdict(lambda: defaultdict(lambda: defaultdict()))
+    )
+    """Cache for storage instances by alias, separated by storage type and additional hash key."""
+
+
 StorageClientType = DatasetClient | KeyValueStoreClient | RequestQueueClient
 """Type alias for the storage client types."""
 
-ClientOpener = Callable[..., Awaitable[StorageClientType]]
+ClientOpener = Coroutine[None, None, StorageClientType]
```

Comment:
ClientOpener - rename it please (all occurrences), as I pointed out somewhere else